--- pdksh-5.2.14/history.c.orig	Wed Jul 18 19:21:08 2001
+++ pdksh-5.2.14/history.c	Wed Jul 18 20:06:23 2001
@@ -714,7 +714,10 @@
 				hline[nread] = '\n';
 			}
 			end = strchr(hline + pos, '\n'); /* will always succeed */
-			*end = '\0';
+			if(end != NULL) /* ...or NOT! (if sth put \0 in .history) */
+				*end = '\0';
+			else
+				end = strchr(hline + pos, '\0'); /* THIS will always succeed */
 			if (contin)
 				histappend(hline + pos, 0);
 			else {
