--- Makefile.in~	Fri Apr 16 21:19:24 1999
+++ Makefile.in	Sat Jun  5 05:38:38 1999
@@ -21,6 +21,7 @@
 LDFLAGS = @LDFLAGS@
 
 SHELL_PROG = @SHELL_PROG@
+SHELL_PROG_STATIC = $(SHELL_PROG).static
 
 prefix = @prefix@
 exec_prefix = @exec_prefix@
@@ -70,7 +71,7 @@
 	os2/emacs.out os2/kshrc.ksh os2/make.sed os2/os2.c os2/os2siglist.out \
 	os2/README.os2 os2/NEWS.os2 os2/os2bugs os2/th.cmd os2/config.cache
 
-all: $(SHELL_PROG)$(exe_suffix) $(SHELL_PROG).1
+all: $(SHELL_PROG)$(exe_suffix) $(SHELL_PROG_STATIC) $(SHELL_PROG).1
 
 # This shouldn't be first - some makes don't know about PRECIOUS and assume it
 # is the default target.
@@ -81,13 +82,8 @@
 
 install: installdirs all
 	$(INSTALL_PROGRAM) $(SHELL_PROG)$(exe_suffix) $(bindir)/`echo $(SHELL_PROG)|sed '$(transform)'`$(exe_suffix)
+	$(INSTALL_PROGRAM) $(SHELL_PROG_STATIC) $(bindir)/`echo $(SHELL_PROG_STATIC)|sed '$(transform)'`
 	-$(INSTALL_DATA) $(SHELL_PROG).1 $(mandir)/`echo $(SHELL_PROG)|sed '$(transform)'`.$(manext)
-	-@prog=$(bindir)/`echo $(SHELL_PROG)|sed '$(transform)'`$(exe_suffix);\
-	    test -f /etc/shells \
-	    && (grep "^$$prog\$$" /etc/shells > /dev/null \
-		|| echo \
-"NOTE: /etc/shells does not contain $$prog \
-      you should add it if you want to set your shell to $(SHELL_PROG)")
 
 installdirs:
 	$(srcdir)/mkinstalldirs $(bindir) $(mandir)
@@ -101,6 +97,9 @@
 
 $(SHELL_PROG)$(exe_suffix): $(OBJS)
 	$(CC) $(LDFLAGS) -o $@ $(OBJS) $(LIBS)
+
+$(SHELL_PROG_STATIC): $(OBJS)
+	$(CC) $(LDFLAGS) -static -o $@ $(OBJS) $(LIBS)
 
 # two steps to prevent the creation of a bogus tmpksh.1
 $(SHELL_PROG).1: $(srcdir)/ksh.Man
