diff -urN pdksh-5.2.14.org/var.c pdksh-5.2.14/var.c
--- pdksh-5.2.14.org/var.c	2004-04-02 14:55:10.447974032 +0200
+++ pdksh-5.2.14/var.c	2004-04-02 14:57:59.620255936 +0200
@@ -439,6 +439,15 @@
 	base = 10;
 	num = 0;
 	neg = 0;
+	if (*s == '0') {
+	    s++;
+	    if (*s == 'x' || *s == 'X') {
+		s++;
+		base = 16;
+	    } else
+		base = 8;
+	    have_base++;
+	}
 	for (c = *s++; c ; c = *s++) {
 		if (c == '-') {
 			neg++;
