--- pdksh-5.2.14/main.c~	Thu May 30 13:00:18 2002
+++ pdksh-5.2.14/main.c	Thu May 30 13:00:18 2002
@@ -47,7 +47,6 @@
 	  "hash=alias -t",	/* not "alias -t --": hash -r needs to work */
 	  "type=whence -v",
 #ifdef JOBS
-	  "stop=kill -STOP",
 	  "suspend=kill -STOP $$",
 #endif
 #ifdef KSH
--- pdksh-5.2.14/ksh.Man~	Thu May 30 13:02:26 2002
+++ pdksh-5.2.14/ksh.Man	Thu May 30 13:02:26 2002
@@ -513,8 +513,6 @@
 .br
 r='fc \-e \-'
 .br
-stop='kill \-STOP'
-.br
 suspend='kill \-STOP $$'
 .ksh)
 .br
